src/core/unit.c

```diff
@@ -438,6 +438,22 @@ void unit_add_to_dbus_queue(Unit *u) {
         u->in_dbus_queue = true;
 }
 
+void unit_add_to_stop_when_unneeded_queue(Unit *u) {
```

Comment:
I'd prefer to have the fact that this is a nop when `stop_when_unneeded` is not set and in other conditions reflected in the name of the function. Current name is rather misleading. Maybe `unit_submit_to_stop_when_unneeded_queue`?